torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -1286,36 +1296,57 @@ def _build_ddp_param_buckets(self) -> None:
-        if ``params`` is valid.
+        as a :class:`list` of all parameters if ``params`` is valid.
 
-        While :class:`optim.Optimizer <torch.optim.Optimizer>` allows
-        ``params`` to be an iterable of :class:`dict` s, currently
-        ``ZeroRedundancyOptimizer`` strictly requires ``params`` to be an
-        iterable of :class:`torch.Tensor` s.
+        Arguments:
+            params (Any): Candidate parameter list or parameter groups to
+                verify.
 
         Raises:
             TypeError: ``params`` has an invalid type.
             ValueError: ``params`` is empty.
+
+        Return:
+            The persistent form of ``params`` to be passed into the parent
+            :class:`Optimizer` constructor -- i.e. returns ``params`` as a
+            :class:`list` if it is originally any iterable of
+            :class:`torch.Tensor` to ensure it can be iterated over again, and
+            returns ``params`` as is if it is an iterable of ``dict``.
         """
         if isinstance(params, torch.Tensor):
-            raise TypeError("params argument should be an iterable of "
+            raise TypeError("`params` argument should be an iterable of "
                             f"Tensors, but got {torch.typename(params)}")
         try:
-            self._all_params = list(params)
+            all_params = list(params)
         except TypeError:
-            raise TypeError("params argument should be an iterable of "
+            raise TypeError("`params` argument should be an iterable of "
                             f"Tensors, but got {torch.typename(params)}")
-        if len(self._all_params) == 0:
+        if len(all_params) == 0:
             raise ValueError("ZeroRedundancyOptimizer got an empty parameter "
                              "list")
-        for param in self._all_params:
-            if not isinstance(param, torch.Tensor):
-                raise TypeError("params argument should be an iterable of "
-                                "Tensors, but got an iterable containing "
-                                f"{torch.typename(param)}")
+        all_tensors = True
+        all_dicts = True
+        for param in all_params:
+            all_tensors &= isinstance(param, torch.Tensor)
+            all_dicts &= isinstance(param, dict)
+        if not all_tensors and not all_dicts:
+            raise TypeError("`params` argument should be an iterable of "
+                            "Tensors or dicts")
+        # Ensure that `self._all_params` contains a list of all parameters
+        if all_tensors:
+            self._all_params = all_params
+            return self._all_params
+        elif all_dicts:
+            self._all_params = []
+            # `all_params` contains parameter groups (not parameters)
+            for param_group in all_params:
+                self._all_params.extend(param_group["params"])
+            return params
```

Comment:
I agree that this is confusing, but hopefully the docstring clarifies things (see the `Returns`).

In the current design, the input is treated as possibly malformed upon entering this method, and `ZeRO` performs its own validation before forwarding arguments to its parent `Optimizer` constructor later in `__init__()`.

The return value of this method is meant to be exactly what to forward into the parent `Optimizer` constructor. The rest of the `ZeRO` implementation relies on `self._all_params` representing a `list` of all parameters (and not any kind of `dict`), so I wanted to preserve that meaning of `self._all_params`.

Now, if `params` is an iterable of `dict`s, then we should maintain that when passing to the parent `Optimizer` constructor to correctly construct `self.param_groups`, so we return `params`. On the other hand, if `params` is an iterable of `Tensor`s, then it may be a generator, in which case it gets exhausted during the validation --- there are some tricky Python semantics at play here, and I could not determine a better solution. Python does not support cloning generators. This is why we return `self._all_params`, which is a `list` and thus can be iterated over again.

Reply:
Update: I realized that if somehow `params` is a generator of `dict`s, the same issue holds, so I have updated the logic to handle that case (which probably never happens in practice but is still permitted under the API definition).

This _should_ make things slightly less confusing as well since we always return `all_params` now.